### PR TITLE
55566 - I have Implemented the engineer Login test And Improved the Project Architecture 

### PR DESCRIPTION
**Pull request by Erlinguss**
image

1. I have implemented the Repository pattern by adding IRepository and MemoryRepository for managing data within the test flow.
2. I have created the BaseStep and IStep abstractions, which define the structure and chaining mechanism for test steps.
3. I have set up the FlowFactory to manage predefined flows, making the flow creation process reusable and organized.
4. I have added Models such as AdminData and EngineerData to encapsulate and manage the test data effectively.
5. I have categorized the Steps into the Login folder, where I implemented NavigateToLogin, FillEngineerDetails, and ClickLoginButton for the Engineer Login flow.
6. I have designed the Assertions to validate the test outcome, specifically creating EngineerLoginAssertion to verify the successful login message.
7. I have organized the test setup and teardown logic in the TestBase, ensuring consistent initialization and cleanup for Playwright tests.
8. I have implemented the test logic directly in the EngineerLoginTests, chaining steps (NavigateToLogin, FillEngineerDetails, ClickLoginButton) and the assertion to create the flow dynamically.
9. I have ensured the Repository is initialized and populated with necessary test data (EngineerData and Page) within the test case.
10. I have handled exceptions during the test execution by adding a try-catch block to log errors gracefully.
11.
Note: Abstractions and Helpers depending on the complexity of the project I will probably implement them as a NuGet packages to make the project cleaner (If needed).

### CODE REVIEW

File: Frontend/UpFluxAutomation/Flows/FlowFactory.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
+﻿using Microsoft.Playwright;
+using System;
+using UpFluxAutomation.Abstractions;
+using UpFluxAutomation.Helpers;
+using UpFluxAutomation.Steps;
+
+namespace UpFluxAutomation.Flows
+{
+    public static class FlowFactory
+    {
```
Comment (PatrickOrjieh) [inline Frontend/UpFluxAutomation/Flows/FlowFactory.cs:10]: Is this flow factory a representation of the Factory Pattern?

File: Frontend/UpFluxAutomationTest/LoginTest/EngineerLoginTest.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+{
+    [TestFixture]
+    public class EngineerLoginTests : TestBase
+    {
+        [Test]
+        public async Task TestEngineerLogin()
+        {
+            try
+            {
+                // Create and initialize EngineerData
```
Comment (PatrickOrjieh) [inline Frontend/UpFluxAutomationTest/LoginTest/EngineerLoginTest.cs:19]: Will this be updated like we discussed to use only token to login?

File: Frontend/UpFluxAutomation/Flows/FlowFactory.cs (hunk 1 above)
Comment (Erlinguss) [reply Frontend/UpFluxAutomation/Flows/FlowFactory.cs:10]: Yes, this FlowFactory represents  the Factory Pattern, as it encapsulates the creation logic for different predefined flows. However, it is designed to dynamically chain and configure test steps specific to this automation framework. You can do the test steps in the main test case but in some scenarios will be redundant repeating the same test steps just for login when you can encapsulate this in the FlowFactory.

File: Frontend/UpFluxAutomationTest/LoginTest/EngineerLoginTest.cs (hunk 2 above)
Comment (Erlinguss) [reply Frontend/UpFluxAutomationTest/LoginTest/EngineerLoginTest.cs:19]: Yes, the test can be updated to use only the token for login. but that's need to be modified in the UI, As I'm testing what is available i need to execute the test like this way, either that, the project is setup to handle everything Dynamically(avoiding dependencies), that is why  takes me a while just to built the logic.

Comment (PatrickOrjieh) [reply Frontend/UpFluxAutomationTest/LoginTest/EngineerLoginTest.cs:19]: Okay, good job

**Review by PatrickOrjieh — APPROVED** [review]

File: Frontend/UpFluxAutomation/Models/EngineerData.cs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+﻿namespace UpFluxAutomation.Helpers
+{
+    public class EngineerData
+    {
+        public string Email { get; set; }
```
Comment (gursimar03) [inline Frontend/UpFluxAutomation/Models/EngineerData.cs:5]: We plan to remove email from engineers api.. so keep in mind to change this in future

**Review by gursimar03 — APPROVED** [review]

File: Frontend/UpFluxAutomation/Models/EngineerData.cs (hunk 3 above)
Comment (Erlinguss) [reply Frontend/UpFluxAutomation/Models/EngineerData.cs:5]: Perfect I keep in mind that.

**Pull request merged**